paddle/fluid/operators/mkldnn/prelu_mkldnn_op.cc

```diff
@@ -0,0 +1,162 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/platform/mkldnn_reuse.h"
+
+namespace paddle {
+namespace operators {
+
+using dnnl::memory;
+using framework::Tensor;
+using platform::GetMKLDNNFormat;
+using platform::MKLDNNDeviceContext;
+using platform::MKLDNNGetDataType;
+using platform::to_void_cast;
+
+template <typename T>
+class PReluMKLDNNHandler
+    : public platform::MKLDNNHandlerT<T, dnnl::prelu_forward,
+                                      dnnl::prelu_backward> {
+ public:
+  PReluMKLDNNHandler(const MKLDNNDeviceContext& dev_ctx,
+                     const mkldnn::engine engine, platform::Place cpu_place,
+                     const Tensor* x, const Tensor* weights,
+                     const std::string& uniq_name, bool is_test = false)
+      : platform::MKLDNNHandlerT<T, dnnl::prelu_forward, dnnl::prelu_backward>(
+            dev_ctx, engine, cpu_place,
+            platform::CreateKey(dev_ctx, framework::vectorize(x->dims()),
+                                uniq_name)) {
+    if (!this->isCached()) {
+      auto x_md = memory::desc(framework::vectorize(x->dims()),
+                               MKLDNNGetDataType<T>(), x->format());
+      auto weights_md =
+          memory::desc(framework::vectorize(weights->dims()),
+                       MKLDNNGetDataType<T>(), memory::format_tag::any);
+
+      this->AcquireForwardPrimitiveDescriptor(dnnl::prop_kind::forward_training,
+                                              x_md, weights_md);
+      if (!is_test)
+        this->AcquireBackwardPrimitiveDescriptor(x_md, weights_md, x_md,
+                                                 weights_md);
+    }
+  }
+
+  std::shared_ptr<memory> AcquireWeightsMemoryWithReorder(const Tensor* input,
+                                                          const bool is_test) {
+    const T* input_data = input->data<T>();
+    auto user_weights_md =
+        memory::desc(framework::vectorize(input->dims()),
+                     MKLDNNGetDataType<T>(), input->format());
+    return this->AcquireMemoryWithReorder(
+        user_weights_md, this->fwd_pd_->weights_desc(),
+        to_void_cast<T>(input_data), "@alpha_mem_p", is_test);
+  }
+
+  std::shared_ptr<memory> AcquireDiffWeightsMemory(Tensor* output) {
+    T* output_data = output->mutable_data<T>(
+        this->place_, this->bwd_pd_->diff_weights_desc().get_size());
+    return this->AcquireMemoryFromPrimitive(this->bwd_pd_->diff_weights_desc(),
+                                            output_data, "@diff_weights_mem_p");
+  }
+};
+
+template <typename T>
+class PReluMKLDNNKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    this->RunKernel(ctx);
+  }
+
+  void RunKernel(const framework::ExecutionContext& ctx) const {
```

Comment:
That's some kind of semantics which I use in every operator. It will be easier for future to add functionality like "force_fp32_output" when we have two different functions for that or if OneDNN will have more support for mixed-precision, but if you want I can merge these two functions